### PR TITLE
FIX Only call getNonce when the method is available on the controller

### PR DESCRIPTION
**Pull request by elliot-sawyer**
When you've configured the module to apply nonces to javascript and CSS, an additional check is needed before applying the nonce to a script, link, or style tag. The codebase often calls `Controller::curr()->getNonce()` but assumes `getNonce` is available on that controller.  This fix ensures `getNonce()` can only be called if the extension has been applied to the controller

### CODE REVIEW

**Comment by Firesphere** [comment]
Could you add some tests for this?

e.g. forcefully unregister the extension before running the test, to confirm no nonce is generated?

**Comment by Firesphere** [comment]
Please ignore the failed test, unless you know of a way on how to fix it :)

**Comment by Firesphere** [comment]
Do you have an example of where this might happen? Given it's extending controller, it would only break if you're using ViewableData directly, which shouldn't touch the extension.

**Comment by elliot-sawyer** [comment]
It was triggering on /Security/login when I was trying to visit /admin. I think that's because by default, the extension is only applied to PageController... while `/Security` descends directly from `Controller`. In this case, `getNonce` would not be available on `Controller::curr()` if `CSPBackend.useNonce` is `true`

I'll have a look at adding the tests and fixing that one. :)

**Comment by Firesphere** [comment]
Right, I think I did that, because extending controller directly would hit the CMS, causing some issues there.

I might want/need to experiment to extend Controller directly but add some rules for the CMS, as the CMS probably will do a lot of unsafe-eval

**Comment by elliot-sawyer** [comment]
Lots of unsafe-eval and unsafe-inline for the CMS, along with violations of font-src. 

`CSPBackend.useNonce` applies to the entire module, regardless of where the extension is applied, so the method would be missing from /Security on default settings.  I think it should definitely be left up to the dev to enable the module for the CMS, but in the meantime, this safely applies the nonce when the option is set, and it's available to be used.

**Comment by elliot-sawyer** [comment]
I've updated the tests to cover this particular issue. Still not sure how to fix the 127.0.0.1 test but I'll look at fixing it in a different PR. :)

**Comment by Firesphere** [comment]
Possibly get a file that's not excluded. A dummy .txt would work?

File: src/Builders/CSSBuilder.php [hunk 1]
```diff
@@ -85,7 +85,10 @@ public function getHeadTags(string &$requirements): void
             $content = array_keys($item)[0];
             $options = $item[$content] ?? [];
             if (CSPBackend::isUsesNonce()) {
-                $options['nonce'] = Controller::curr()->getNonce();
+                $ctrl = Controller::curr();
```
Comment (Firesphere) [inline src/Builders/CSSBuilder.php]: Could this be resolved by `Controller::hasCurr()`?

File: src/Builders/CSSBuilder.php [hunk 2]
```diff
@@ -108,7 +111,10 @@ public function getCustomTags(string $requirements): string
             $options = ['type' => 'text/css'];
             // Use nonces for inlines if requested
             if (CSPBackend::isUsesNonce()) {
-                $options['nonce'] = Controller::curr()->getNonce();
+                $ctrl = Controller::curr();
```
Comment (Firesphere) [inline src/Builders/CSSBuilder.php]: As above, and same for as below

File: tests/unit/CSPBackendTest.php [hunk 3]
```diff
@@ -6,12 +6,18 @@
 use Firesphere\CSPHeaders\Builders\JSBuilder;
 use Firesphere\CSPHeaders\Extensions\ControllerCSPExtension;
 use Firesphere\CSPHeaders\View\CSPBackend;
+use ParagonIE\Halite\Symmetric\Config;
```
Comment (Firesphere) [inline tests/unit/CSPBackendTest.php]: Eh?

File: src/Builders/CSSBuilder.php (hunk 1 above)
Comment (elliot-sawyer) [reply src/Builders/CSSBuilder.php]: The issue is that `getNonce()` is applied globally in config, but the extension isn't applied to the CMS controllers. I've added a fix for `Controller::has_curr()`, but it does still need to check if the method exists.

File: src/Builders/CSSBuilder.php (hunk 2 above)
Comment (elliot-sawyer) [reply src/Builders/CSSBuilder.php]: as above and below

File: tests/unit/CSPBackendTest.php (hunk 3 above)
Comment (elliot-sawyer) [reply tests/unit/CSPBackendTest.php]: fixed!

File: src/Builders/CSSBuilder.php [hunk 4]
```diff
@@ -107,8 +110,11 @@ public function getCustomTags(string $requirements): string
         foreach ($this->getOwner()->getCustomCSS() as $css) {
             $options = ['type' => 'text/css'];
             // Use nonces for inlines if requested
-            if (CSPBackend::isUsesNonce()) {
-                $options['nonce'] = Controller::curr()->getNonce();
+            if (CSPBackend::isUsesNonce() && Controller::has_curr()) {
```
Comment (Firesphere) [inline src/Builders/CSSBuilder.php]: I think the `has_curr()` here is redundant?

**Comment by phalkunz** [comment]
I ran into the same issue here - navigate to `Security/login`. Anyways chance of merging soon?

**Comment by Firesphere** [comment]
@elliot-sawyer Think you can fix up the PHPCS issues? Then I can merge, coverage can be fixed up later

**Comment by Firesphere** [comment]
@elliot-sawyer @phalkunz The amount of errors and code coverage, has me a bit worried, so I still haven't merged it.

@elliot-sawyer Can you please go over the errors of codeclimate/codecov/CircleCI and fix them up? It shouldn't be too much work, but I really don't have time to look at it at the moment.

**Comment by Firesphere** [comment]
@phalkunz The only short-term issue I can think of, is adding an extension to the Security controller, and in the constructor or init, remove the CSP extension.

**Pull request merged**